proposals/0021-vk-coop-matrix.md

```diff
@@ -0,0 +1,466 @@
+to match the corresponding operations in the
+(SPV_KHR_cooperative_matrix)[https://htmlpreview.github.io/?https://github.com/KhronosGroup/SPIRV-Registry/blob/main/extensions/KHR/SPV_KHR_cooperative_matrix.html]
+extension.
+
+To help implement `vk::khr::CooperativeMatrix`, we will also introduce utility
+classes `vk::util::ArithmeticSelector` and `vk::util::ConversionSelector`. This
+class will be defined in a header file "vk/opcode_selector.h". They can be used
+to generate inline SPIR-V arithmetic and conversion instructions with the
+correct opcode for the types.
+
+The interface for `vk::khr::CooperativeMatrix` uses enums that are defined in
+the SPIR-V specification. The required enums will be defined in "vk/spirv.h".
+
+## Detailed design
+
+### `vk::khr::CooperativeMatrix`
+
+The `CooperativeMatrix` class will be defined in a header file as a wrapper
+around a `vk::SpirvType` that will expand to the appropriate SPIR-V type. This
+class will have the following interface.
+
+```c++
+
+// The base cooperative matrix class. The template arguments correspond to the
+// operands in the OpTypeCooperativeMatrixKHR instruction.
+template <typename ComponentType, Scope scope, uint rows, uint columns,
+          CooperativeMatrixUse use>
+class CooperativeMatrix {
+  template <class NewComponentType>
+  CooperativeMatrix<NewComponentType, scope, rows, columns, use> cast();
+
+  // Apply OpSNegate or OFNegate, depending on ComponentType, in a element by
+  // element manner.
+  CooperativeMatrix negate();
+
+  // Apply OpIAdd or OFAdd, depending on ComponentType, in a element by element
+  // manner.
+  CooperativeMatrix operator+(CooperativeMatrix other);
+
+  // Apply OpISub or OFSub, depending on ComponentType, in a element by element
+  // manner.
+  CooperativeMatrix operator-(CooperativeMatrix other);
+
+  // Apply OpIMul or OFMul, depending on ComponentType, in a element by element
+  // manner.
+  CooperativeMatrix operator*(CooperativeMatrix other);
+
+  // Apply OpSDiv, OpUDiv or OFDiv, depending on ComponentType, in a element by
+  // element manner.
+  CooperativeMatrix operator/(CooperativeMatrix other);
+
+  // Apply OpMatrixTimesScalar in a element by element manner.
+  CooperativeMatrix operator*(ComponentType scalar);
+
+  // Store the cooperative matrix using OpCooperativeMatrixStoreKHR to
+  // data using the given memory layout, stride, and memory access mask.
+  // `NonPrivatePointer` and `MakePointerVisible` with the workgroup scope
+  // will be added to the memory access mask to make the memory coherent.
+  //
+  // This function uses a SPIR-V pointer because HLSL does not allow groupshared
+  // memory object to be passed by reference. The pointer is a hack to get
+  // around that.
+  //
+  // The layout and stride will be passed to the SPIR-V instruction as is. The precise meaning can be found in the specification for SPV_KHR_cooperative_matrix.
+  template <MemoryAccessMask memoryAccessMask, CooperativeMatrixLayout layout,
```

Comment:
sort of a nitpick, but the strongly typed MemoryAccessMask enum will be clunky (and technically undefined behavior) if the shader wants to use more than one bit. I wonder if we still even need memoryAccessMask after how coherent has been resolved?

Reply:
I kept the version with the memory access mask just. There are a couple things people could do with it, even if it is not common. The extra testing is not that onerous. It will also future proof a little.

Yes the use of the bit mask is clunky, but I expect the overloads without the memory access mask to be used most of the time. It is only ugly for those who need to so something special.

I change the type to uint32_t to avoid the undefined behaviour.